rholang-tutorial/src/components/Markdown.js

```diff
@@ -0,0 +1,62 @@
+import React from 'react'
+import ReactHtmlParser, { convertNodeToElement, processNodes } from 'react-html-parser'
+import generatePropsFromAttributes from 'react-html-parser/lib/utils/generatePropsFromAttributes'
+import compiler from 'react-smackdown/lib/compiler'
+
+import Code from './Code'
+import SubHeading from './SubHeading'
+
+const renderers = {
+  'h2': (props) => (<SubHeading level={2} {...props}/>),
+  'h3': (props) => (<SubHeading level={3} {...props}/>)
+}
+
+let setToc = false
+
+export default function Markdown ({source, onTOC = () => {}}) {
+  let toc = []
+  const transform = (node, index) => {
+    // Transform any component renderer as react component
+    if (node.type === 'tag' && renderers[node.name]) {
+      const Component = renderers[node.name]
+      const props = generatePropsFromAttributes(node.attribs, index)
+      const children = processNodes(node.children, transform)
+
+      const flattenedChildren = node.children && node.children[0] && node.children[0].data
+        ? node.children[0].data : null
+
+      if(!setToc) {
+        toc.push({type: node.name, id: props.id, value: flattenedChildren})
+      }
+
+      return React.createElement(Component, props, children)
+    }
+
+    // Transform any code examples
+    if (node.type === 'tag' && node.name === 'pre') {
```

Comment:
Only code samples with line numbers should be converted, so include a check that /^\s*\d+/ matches the first line.

Reply:
Actually, I think all code samples should be converted from boring `<pre>`'s to their syntax highlighted counterparts.

There is already code present to check whether a 'Run' button should be shown depending on whether there are any line numbers in the sample: https://github.com/rchain/rchain/pull/683/files#diff-a59c22e3ffbd92939dc8f2d14b61c6c1R19

(related issue from earlier today: https://github.com/tschoffelen/rholang-tutorial/issues/1)

Reply:
Thanks for checking this so thoroughly by the way! @pyrocto

Reply:
Sweet! LGTM.